### PR TITLE
Everything

### PR DESCRIPTION
**Pull request by Avaq**
**[View README](https://github.com/fluture-js/fluture-retry)**

### CODE REVIEW

**Comment by Avaq** [comment]
I also like the version proposed by @semmel in https://github.com/fluture-js/Fluture/issues/219, where the user can decide which specific errors to retry.

The reason I hadn't published fluture-retry yet, even though it's been sitting on my machine for almost a year, is because there are so many different ways one can implement retries. I kept adding on new parts (like logging in between tries) and then removing them again for other use-cases. I could not reach a consensus with *myself* about the features that should be included. This PR contains a stripped down version.

File: index.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
+//. # Fluture retry
+//.
+//. Toolset for retrying potentially failing computations represented by
+//. [Fluture][] Futures.
+//.
+//. ```sh
+//. npm install --save fluture fluture-retry
+//. ```
```
Comment (davidchambers) [inline index.js]: Another option:

    ```console
    $ npm install --save fluture fluture-retry
    ```

File: index.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+}(function(Future) {
+
+  'use strict';
+
+  //  last :: NonEmpty (Array a) -> a
+  function last(xs) {
+    return xs[xs.length - 1];
+  }
+
+  //# retry :: retry :: (Number -> Number, Number, Future a b) -> Future (Array a) b
```
Comment (davidchambers) [inline index.js]:
```diff
-  //# retry :: retry :: (Number -> Number, Number, Future a b) -> Future (Array a) b
+  //# retry :: (Number -> Number, Number, Future a b) -> Future (Array a) b
```

File: index.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  //.
+  //. Create a retrying Future using the given parameters:
+  //.
+  //. 1. A function over the amount of failures to determine waiting time.
+  //.    See [`exponentially`](#exponentially), [`linearly`](#linearly) and
+  //.    [`statically`](#statically) for pre-baked functions of this sort.
+  //. 1. A natural number representing the maximum amount of retries.
```
Comment (davidchambers) [inline index.js]: I would replace _amount_ with _number_ in this sentence.

File: test/index.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
+'use strict';
+
+var assert = require('assert');
+var Z = require('sanctuary-type-classes');
+var Future = require('fluture');
```
Comment (davidchambers) [inline test/index.js]: :abc: ?

**Comment by Avaq** [comment]
Thank you @davidchambers :)

File: .config [hunk 5]
```diff
@@ -0,0 +1,2 @@
+repo-owner = fluture-js
+repo-name = fluture-retry
```
Comment (davidchambers) [inline .config:3]: You should add this line:

    author-name = Aldwin Vlasblom

File: package.json [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  "author": "Aldwin Vlasblom <aldwin.vlasblom@gmail.com> (https://github.com/Avaq)",
+  "license": "MIT",
+  "dependencies": {},
+  "peerDependencies": {
+    "fluture": ">=6.0.0 <9.0.0"
+  },
+  "devDependencies": {
+    "codecov": "^3.0.0",
+    "fluture": "^8.0.1",
+    "sanctuary-scripts": "^1.0.1"
```
Comment (davidchambers) [inline package.json:39]: Do you want `"^1.0.1"` or `"~1.0.1"` here?

Comment (Avaq) [reply package.json:39]: I believe `^` includes minor and patch versions. The way I see it: modules usually get better over time, so as long as the developer doesn't indicate a breaking change, I'm interested in the updates.

How do you view it, David?

Comment (davidchambers) [reply package.json:39]: My habit is to use `"1.0.x"`. My reasoning is that if v1.1.0 is released with a new feature I wish to use, I'll need to change the specification to `"1.1.x"` (`"1.x.x"` would not suffice). Your approach is equally valid. In this scenario you would change the specification from `"^1.0.1"` to `"^1.1.0"`.

In summary, I should update my mental model. :)

**Review by davidchambers — APPROVED** [review]

**Comment by semmel** [comment]
I am impressed that it just took a few hours from my idea of a retry function to this nice little extension library. Great stuff!

I' d just like to point out two things from my user point of view:
* In general I tend to hold the number of dependencies in my projects down. So while using retry this way would be perfectly easy, on the other hand I am not convinced that it justifies to depend on yet another library. I am aware that there is even [node-retry](https://github.com/tim-kos/node-retry) which makes a big science (not meant in a bad way) of retry strategies. In real life however I would just put those formulas in-line as an argument for retry.
* User requirements for retrying stuff can be very diverse. My current use case is to open a Websocket connection to a server: Depending on the error I'd choose a different retry strategy: 
    * If the device has no connection to the internet, linearly,
    * if our server is too busy, exponentially to avoid a DOS attack

Maybe I still havn't thought that to the end, but that last problem I could just solve in user-land by hacking the code for `retry` from the library.

**Comment by Avaq** [comment]
> User requirements for retrying stuff can be very diverse -- @semmel 

That's exactly the curse of this library. It's too specific, and if I make it more generic it quickly becomes *too generic*. Too generic means that it requires the user to type just about as much code as they would have without the library.

Like you said - I'll likely be writing custom retry logic for my specific use-cases and not even use the library in many cases. But since it's not doing anyone any good collecting dust, I'm putting it out there. Hopefully it will do someone some good, if only to inspire them.

**Pull request merged**